op-node/sources/engine_client.go

```diff
@@ -0,0 +1,110 @@
+package sources
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/ethereum-optimism/optimism/op-node/client"
+	"github.com/ethereum-optimism/optimism/op-node/eth"
+	"github.com/ethereum-optimism/optimism/op-node/rollup"
+	"github.com/ethereum-optimism/optimism/op-node/sources/caching"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/rpc"
+)
+
+type EngineClientConfig struct {
+	L2ClientConfig
+}
+
+func EngineClientDefaultConfig(config *rollup.Config) *EngineClientConfig {
+	return &EngineClientConfig{
+		// engine is trusted, no need to recompute responses etc.
+		L2ClientConfig: *L2ClientDefaultConfig(config, true),
+	}
+}
+
+// EngineClient extends L2Client with engine API bindings.
+type EngineClient struct {
+	*L2Client
+}
+
+func NewEngineClient(client client.RPC, log log.Logger, metrics caching.Metrics, config *EngineClientConfig) (*EngineClient, error) {
+	l2Client, err := NewL2Client(client, log, metrics, &config.L2ClientConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	return &EngineClient{
+		L2Client: l2Client,
+	}, nil
+}
+
+// ForkchoiceUpdate updates the forkchoice on the execution client. If attributes is not nil, the engine client will also begin building a block
+// based on attributes after the new head block and return the payload ID.
+// May return an error in ForkChoiceResult, but the error is marshalled into the error return
+func (s *EngineClient) ForkchoiceUpdate(ctx context.Context, fc *eth.ForkchoiceState, attributes *eth.PayloadAttributes) (*eth.ForkchoiceUpdatedResult, error) {
+	e := s.log.New("state", fc, "attr", attributes)
+	e.Trace("Sharing forkchoice-updated signal")
+	fcCtx, cancel := context.WithTimeout(ctx, time.Second*5)
+	defer cancel()
+	var result eth.ForkchoiceUpdatedResult
+	err := s.client.CallContext(fcCtx, &result, "engine_forkchoiceUpdatedV1", fc, attributes)
+	if err == nil {
+		e.Trace("Shared forkchoice-updated signal")
+		if attributes != nil {
```

Comment:
Why does this check attributes here but then reference result? It looks like this if not nil check is to prevent referencing a nil pointer

Reply:
`attributes` is an optional argument, not all forkchoice updates initiate block building. I'll add a comment